File: 2022/day-8/day8.py

import numpy as np
from typing import *


def parse_input(inp: str) -> np.ndarray:
    return np.array([list(map(int, line)) for line in inp.splitlines()])


def get_directions(i: int, j: int, grid: np.ndarray) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
    return grid[i, :j][::-1], grid[i, j+1:], grid[:i, j][::-1], grid[i+1:, j]


def find_visible(grid: np.ndarray) -> Iterable[Tuple[int, int, int]]:
    for (i, j), height in np.ndenumerate(grid):
        if any(np.all(direction < height) for direction in get_directions(i, j, grid)):
            yield i, j, height


def axis_score(height: int, ax: np.ndarray) -> int:
    if not ax.size:
        return 0
    score = 1
    for other_height in ax[:-1]:
        if height > other_height:
            score += 1
        else:
            break
    return score


def compute_view_score(grid: np.ndarray, visible: Iterable[Tuple[int, int, int]]) -> Iterable[int]:
    for i, j, height in visible:
        dirs = get_directions(i, j, grid)
        axis_scores = map(lambda arr: axis_score(height, arr), dirs)
        yield np.prod(list(axis_scores))


with open('input.txt') as h:
    board = parse_input(h.read())

print('Part 1:', sum(1 for _ in find_visible(board)))
print('Part 2:', max(compute_view_score(board, find_visible(board))))

File: 2022/day-8/day8.rs

type Board = Vec<Vec<u8>>;

fn parse_input(inp: &str) -> Board {
    inp.lines()
        .map(|l| l.chars().map(|c| c.to_digit(10).unwrap() as u8).collect())
        .collect()
}

fn shape(board: &Board) -> (usize, usize) {
    (board.len(), board.get(0).unwrap().len())
}

fn generate_axes(
    i: usize,
    j: usize,
    board: &Board,
) -> (
    impl Iterator<Item = u8> + '_ + Clone,
    impl Iterator<Item = u8> + '_ + Clone,
    impl Iterator<Item = u8> + '_ + Clone,
    impl Iterator<Item = u8> + '_ + Clone,
) {
    let (nx, ny) = shape(board);
    let left = (0..j).rev().map(move |jj| board[i][jj]);
    let right = ((j + 1)..nx).map(move |jj| board[i][jj]);
    let up = (0..i).rev().map(move |ii| board[ii][j]);
    let down = ((i + 1)..ny).map(move |ii| board[ii][j]);
    (left, right, up, down)
}

fn visible_trees(board: &Board) -> impl Iterator<Item = (usize, usize)> + '_ {
    let (nx, ny) = shape(board);
    (0..nx)
        .flat_map(move |i| (0..ny).map(move |j| (i, j)))
        .filter(|&(i, j)| {
            let height = board[i][j];
            let lower = |h| h < height;
            let (mut l, mut r, mut u, mut d) = generate_axes(i, j, board);
            l.all(lower) || r.all(lower) || u.all(lower) || d.all(lower)
        })
}

fn part_one(board: &Board) {
    println!("Part 1: {}", visible_trees(board).count());
}

fn score(ax: impl Iterator<Item = u8> + Clone, height: u8) -> usize {
    let n = ax.clone().count();
    if n == 0 {
        return 0;
    }
    let mut score = 1;
    for other_height in ax.take(n - 1) {
        if height > other_height {
            score += 1;
        } else {
            break;
        }
    }
    score
}

fn part_two(board: &Board) {
    let best_score = visible_trees(board)
        .map(|(i, j)| {
            let h = board[i][j];
            let (l, r, u, d) = generate_axes(i, j, board);
            score(l, h) * score(r, h) * score(u, h) * score(d, h)
        })
        .max();
    println!("Part 2: {}", best_score.unwrap());
}

fn main() {
    let contents = std::fs::read_to_string("input.txt").expect("Failed to read input");
    let board = parse_input(&contents);
    part_one(&board);
    part_two(&board);
}
